src/Volume/EncryptionModeXTS.cpp

```diff
@@ -374,7 +380,11 @@ namespace VeraCrypt
 
 		FAST_ERASE64 (whiteningValue, sizeof (whiteningValue));
 		FAST_ERASE64 (whiteningValues, sizeof (whiteningValues));
-	}
+	#else
+                cipher.SetKeyXTS (SecondaryKey.GetRange (0, cipher.GetKeySize()));
```

Comment:
why we need to update XTS key of the cipher on every DecryptBufferXTS call?
Existing classes already implement appropriate logic of setting keys at correct moment.

Reply:
While this not ideal, I thought it to be the best way to keep the API calls the same. In wolfCrypt, AES-XTS is one object and does not have the abstraction of AES and XTS as two layers. The SetKey system that is currently in place will load the XTS key into the secondary cipher, but with wolfCrypt the primary cipher (AES-XTS) will also need the XTS key. If your preference is to add a call to SetKeyXTS to each EA usage, then I'm sure that can be done

Reply:
Thank you for elaborating on your rationale behind the changes.

Considering the unique abstraction of wolfCrypt, it would seem more maintainable and clean to encapsulate this specific behavior. I propose creating a dedicated class, perhaps named EncryptionModeWolfcryptXTS, that inherits from EncryptionMode. When WOLFCRYPT_BACKEND is defined, VeraCrypt's code can then instantiate this new class instead of the general EncryptionModeXTS.

This design would grant you the flexibility to fully accommodate wolfCrypt's XTS logic without affecting the current architecture. What are your thoughts on this approach?

Reply:
More than happy to make that change. A couple of thoughts to note. EncryptionModeWolfcryptXTS  would be entirely redundant given EncryptionModeXTS exists and the only change would be the call to SetKeyXTS, which doesn't necessarily affect current architecture if WOLFCRYPT_BACKEND is not defined. This will also expand the changes beyond the crypto layer and result in a bigger diff. If the bigger diff is not an issue in the first place, I could resort to the original approach but remove the call to the SetKeyXTS in the Encrypt/Decrypt and just call that function everywhere that EncryptionModeXTS is leveraged (when WOLFCRYPT_BACKEND is defined). Let me know which route you think is best for your product.

Reply:
Thank you for your insights.

I understand your concerns regarding redundancy. My main aim is to preserve the clarity and coherence of VeraCrypt architecture, as this tends to serve well in long-term maintenance. In light of this, I'd propose that EncryptionModeWolfcryptXTS inherit directly from EncryptionModeXTS. This would allow us to isolate the wolfCrypt-specific logic and reduce redundancy, while keeping our architecture consistent.

A bigger diff is not a primary concern for me. Implementing this in a manner that ensures future modifications or integrations are straightforward is essential.

Please proceed with this approach and let me know if you encounter any hurdles along the way.

Reply:
Thanks for the guidance @idrassi. I've made the suggested changes